server/src/message/mod.rs

```diff
@@ -0,0 +1,667 @@
+                                                 speed,
+                                                 size,
+                                                 bullet_speed,
+                                                 bullet_size);
+                "welcome"
+            }
+            &Message::GoAway{ref reason} => {
+                add_data_entry(&mut values, "reason", &reason);
+                "go_away"
+            }
+            &Message::PlayerJoined{id} => {
+                add_data_entry(&mut values, "id", &id);
+                "player_joined"
+            }
+            &Message::PlayerLeft{id} => {
+                add_data_entry(&mut values, "id", &id);
+                "player_left"
+            }
+            &Message::ShotsFired{id, bullet_id, x, y, aim_x, aim_y} => {
+                add_shot_data_entries(&mut values, id, bullet_id, x, y, aim_x, aim_y);
+                "shots_fired"
+            }
+            &Message::PlayerSpawned{id, x, y} => {
+                add_data_id_pos_entries(&mut values, id, x, y);
+                "player_spawned"
+            }
+            &Message::PlayerDestroyed{id, killer_id, bullet_id} => {
+                add_data_entry(&mut values, "id", &id);
+                match (killer_id, bullet_id) {
+                    (Some(killer_id), Some(bullet_id)) => {
+                        add_data_entry(&mut values, "killer_id", &killer_id);
+                        add_data_entry(&mut values, "bullet_id", &bullet_id);
+                    }
+                    (None, None) => {}
+                    _ => panic!("killer_id and bullet_id must be either both Some or both None"),
+                }
+                "player_destroyed"
+            }
+            &Message::PlayerMoving{id, x, y, move_x, move_y} => {
+                add_data_id_pos_moves_entries(&mut values, id, x, y, move_x, move_y);
+                "player_moving"
+            }
+            &Message::PlayerStopped{id, x, y} => {
+                add_data_id_pos_entries(&mut values, id, x, y);
+                "player_stopped"
+            }
+            &Message::WorldState => "world_state",  //TODO
+            &Message::StartMoving{move_x, move_y} => {
+                add_data_move_entries(&mut values, move_x, move_y);
+                "start_moving"
+            }
+            &Message::StopMoving => "stop_moving",
+            &Message::Fire{move_x, move_y} => {
+                add_data_move_entries(&mut values, move_x, move_y);
+                "fire"
+            }
+        };
+
+        let mut root_obj = BTreeMap::new();
+        let _ = root_obj.insert("type".to_string(),
+                                serde_json::Value::String(msg_type.to_string()));
+        if !values.is_empty() {
+            let _ = root_obj.insert("data".to_string(), serde_json::Value::Object(values));
+        }
+
+        serde_json::to_string(&serde_json::Value::Object(root_obj)).unwrap()
+    }
+}
+
+impl FromStr for Message {
```

Comment:
Again, urgh, such monstrously long methods. Every type has to be handled it's own way... why cram it all together like this?